### PR TITLE
Fixed memory leak for closed actionChannels

### PR DESCRIPTION
**Pull request by Andarist**
I don't think this can be tested easily - from the tests we can't check if there is a leak or not as it is observable only internally.

### CODE REVIEW

File: packages/core/src/internal/channel.js [hunk 1]
```diff
@@ -183,14 +194,20 @@ export function multicastChannel() {
       }
 
       const takers = (currentTakers = nextTakers)
-      takers.forEach(taker => {
+
+      for (let i = 0, len = takers.length; i < len; i++) {
```
Comment (Andarist) [inline packages/core/src/internal/channel.js]: this is a hot path, so a for loop should be better here, perf-wise

Comment (shinima) [reply packages/core/src/internal/channel.js]: ok, then let's add a comment that the for-loop is intentional

File: packages/core/src/internal/channel.js [hunk 2]
```diff
@@ -116,15 +121,15 @@ export function eventChannel(subscribe, buffer = buffers.none()) {
 
   unsubscribe = subscribe(input => {
     if (isEnd(input)) {
-      close()
       closed = true
```
Comment (shinima) [inline packages/core/src/internal/channel.js]: How about moving this statement (`closed = true`) inside `close()`?  

`multicastChannel()` and `channel()` have this statement inside their `close()` too.

File: packages/core/src/internal/channel.js [hunk 3]
```diff
@@ -80,17 +80,22 @@ export function channel(buffer = buffers.expanding()) {
...
-          const taker = arr[i]
-          taker(END)
-        }
-      }
+    if (process.env.NODE_ENV !== 'production') {
+      checkForbiddenStates()
+    }
+
+    if (closed) {
+      return
```
Comment (shinima) [inline packages/core/src/internal/channel.js]: 👍  returning early makes code clearer

**Comment by restrry** [comment]
@Andarist I thought that jest could detect memory leaks. doesn't it?

**Comment by Andarist** [comment]
no idea, can't find it in their docs, not sure how reliable the feature would be - especially for case like this, this was rather small memory leak

**Pull request merged**